### PR TITLE
refactor(engine): introduce a `Lookup_alias.t` type in Action_builder

### PR DESCRIPTION
**Pull request by anmonteiro**
- this is a follow-up to #7334, which only changed `dune_rules`
- the 1st commit refactors `Action_builder` to introduce a new intermediate `Lookup_alias` module that can be used in the `Alias_rec` traversal callback
  - the additional information provided, `allowed_build_only_subdirs`, lets the caller know about build-only subdirectories where rules are defined that dune may need to traverse in order to look up alias rules
- the 2nd commit demonstrates how this new type can be used to refactor the recursive aliases code in `dune_rules`.

### CODE REVIEW

File: src/dune_engine/action_builder.mli [hunk 1]
```diff
@@ -113,7 +113,14 @@ end
 
 val alias : Alias.t -> unit t
 
-val dep_on_alias_if_exists : Alias.t -> Alias_status.t t
+module Lookup_alias : sig
+  type t =
+    { alias_status : Alias_status.t
+    ; allowed_build_only_subdirs : Filename.Set.t
```
Comment (snowleopard) [inline src/dune_engine/action_builder.mli:119]: Could you add some comments here explaining this new field?

**Review by snowleopard — APPROVED** [review]
Thanks! I think the change makes sense. It seems worth adding some comments for the new richer API.

@rgrinberg Please take a look too.

**Comment by snowleopard** [comment]
@anmonteiro Thanks for adding the comment. After reading it, the naming started to look a bit odd though: we're calling `dep_on_alias_if_exists` but get not just an alias but some seemingly unrelated information about the build-only directories. Not a very big deal, but perhaps you have any ideas on how to make the naming a little clearer?

**Comment by anmonteiro** [comment]
Good feedback. I'll think of something.

**Comment by anmonteiro** [comment]
I had dropped the ball on this. Just pushed a commit:

- renamed `Lookup_alias.t` to `Alias_build_info`
- renamed `dep_on_alias_if_exists` to `dep_on_alias_build_info_if_exists`.

**Comment by anmonteiro** [comment]
I plan on merging this change once CI passes, given:

- PR has been approved before
- the new naming is strictly better than what I had came up with previously

**Comment by rgrinberg** [comment]
Let me have a look at it tomorrow first. I forgot I was supposed to read this.

File: src/dune_engine/action_builder.ml [hunk 2]
```diff
@@ -278,30 +278,59 @@ module Alias_status = struct
...
+    in
+    { alias_status = status; allowed_build_only_subdirs }
+end
+
+let dep_on_alias_build_info_if_exists alias =
   of_thunk
     { f =
         (fun mode ->
           let open Memo.O in
-          let* definition = Load_rules.alias_exists alias in
```
Comment (rgrinberg) [inline src/dune_engine/action_builder.ml:286]: Do we still need `alias_exists` in the API of `Load_rules`?

**Review by rgrinberg — APPROVED** [review]

**Pull request merged**